File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    
    
    
    int find(vector<int>&root,int x){
        
            while(root[x]==x){
                
                
                    return x;
                
            }
        
            return root[x]=find(root,root[x]);
    }
    
    
    
    void unionSet(vector<int>&root,vector<int>&rank,int x,int y){
        
        
            int rootx=find(root,x);
            int rooty=find(root,y);
            if(rootx!=rooty){
                
                        if(rank[x]>=rank[y]){
                                
                                rank[x]+=rank[y];
                                root[rooty]=rootx;
                            
                        }
                        else{
                            
                                rank[y]+=rank[x];
                                root[rootx]=rooty;
                            
                            
                        }
                
            }
        
    }
    
    int findCircleNum(vector<vector<int>>& isConnected) {
    
            int n=isConnected.size();
            vector<int>root(n,0);
            vector<int>rank(n,1);
            for(int i=0;i<root.size();i++){
                
                        root[i]=i;
                
            }
            
        
            for(int i=0;i<n;i++){
                for(int j=0;j<i;j++){
                        if(isConnected[i][j]==1){
                                    unionSet(root,rank,i,j);
                        }  
                }
            }
        
        
            for(int i=0;i<root.size();i++){
                
                        root[i]=find(root,root[i]); 
                
            }
        
            set<int>s(root.begin(),root.end());
            return s.size();
    }
};

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    
    def find(self,parent,i):
        
        if parent[i]==i:
            return i
        
        parent[i]=self.find(parent,parent[i])
        return parent[i]
    
    def unionset(self,rank,parent,i,j):
        
        parent_i=self.find(parent,i)
        parent_j=self.find(parent,j)
        if parent_i>parent_j:   
                rank[parent_i]+= rank[parent_j]
                parent[parent_j]=parent[parent_i]
        
        else:
                rank[parent_j]+= rank[parent_i]
                parent[parent_j]=parent[parent_i]
        
        
        
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
    
        
        rank=[-1 for i in range(len(isConnected))]
        parent=[i for i in range(len(isConnected))]
                            
                            
        for i in range(len(isConnected)):
            for j in range(len(isConnected[i])):
                    
                    if isConnected[i][j]==1 and i!=j:
                            self.unionset(rank,parent,i,j)
                            print(i,j)
        
           
    
        visited=[]
        for i in range(len(parent)):
                parent[i]=self.find(parent,i)
                
        return len(set(parent))
        
        
            
        
            